### ISSUE TITLE
Nested navigators reset - Drawer doesnt close automatically?

### ISSUE THREAD

**Issue by bonbonio** [react-navigation/react-navigation#4198]
Hello,

Finally made it work so that I can use a switchNavigator to switch between stack A and stack B, but after the switch to stack B, when I click a page in the DrawerNavigator it changes the view, but the drawer does not automatically close. Also if I click a route in the drawer thats already active/open the drawer will not close either.

Currently as a hack, I am calling `this.props.navigation.closeDrawer();` inside the constructor for all the views in Stack B, but thats a horrid approach.

I am using nested stacks, and using react-navigation 2.0.1

**Comment by orette** [comment]
Hey @bonbonio ,

I'm having a somewhat similar problem. I might end up using your hack until I know for sure if this is a bug or i'm going about it the wrong way. 
[I created a snack with my current navigation setup](https://snack.expo.io/HkD9HDfRG)

## Expected Behavior
I expect the functions `openDrawer()` and `closeDrawer()` to be available on the `navigation` object passed into DrawerNavigator' content component.

```
const MainDrawerNavigator = createDrawerNavigator(
  {
    Tabs: MainTabNavigator,
  },
  {
    contentComponent: props => <SideMenu {...props} />,
  }
);
```

## Actual Behavior
I receive the error `undefined is not a function (evaluating this.props.navigation.closeDrawer())` whenever i try to access the function.

## Steps to Reproduce the Problem
[Snack illustrating the issue.](https://snack.expo.io/HkD9HDfRG) 

## package.json
```  
    "expo": "^27.0.0",
    "react": "16.3.1",
    "react-native": "0.55.*",
    "react-navigation": "^2.0.1",
```

**Comment by bonbonio** [comment]
Stack A and B both belong to a root stack C. A + C are stack navigators, and B is a DrawerNavigator

Stack A is my unauthorised routes, B is my authorised roots, c is what is placed in the App.js, so on login I navigate to Stack B from stack A.

All the routes which are not closing the drawer are in stack B.

I'd post code but it would require a bit of pulling apart, but I could do so if its important.

**Comment by orette** [comment]
@bonbonio 

I modified the illustration i did to match your hierarchy, but it seems to be working. Take a look to see if i missed something. [Code Matching your hierarchy.](https://snack.expo.io/@orette/nested-navigation-test)

**Comment by bonbonio** [comment]
@orette my navigation is quite complex, is there another method I can talk to you, to get a pair of eyes on my code thats less public whilst we investigate?

Many Thanks

**Comment by orette** [comment]
@bonbonio I don't mind using your preference. Just let me know what you prefer, and I'll have it setup so we can check it out later.

**Comment by stychu** [comment]
@bonbonio  @orette I have the same problem guys. Fighting this for 3 days already. Maybe could you help me with this? This is my current implementation. When I open the drawer and navigate to the drawer navigation route the drawer doesn't close itself.

App.js
```
const StackNav = createStackNavigator(
  {
    Home: { screen: HomeScreen },
    ScanCode: { screen: ScanCodeModal },
    ScanResult: { screen: ScanResultScreen },
  }, {
    headerMode: 'none'
  });

const Navigator = createDrawerNavigator(
  {
    SignInScreen: { screen: SignInScreen },
    Main: {
      screen: StackNav, navigationOptions: {
        drawerLabel: () => null
      }
    },
    Home: { screen: HomeScreen },
    ScanCode: { screen: ScanCodeModal }
  }, {
    initialRouteName: 'SignInScreen',
    // contentComponent: ({ navigation }) => <Drawer navigation={navigation}/>
  });
```
And in sigin route i reset to:

```
class SignIn extends React.Component {
  static navigationOptions = () => ({
    navBarHidden: true,
    header: null,
    drawerLabel: () => null
  });

  routeToMain(routeName) {
    const resetAction = StackActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName })],
    });
    this.props.navigation.dispatch(resetAction);
  }

  async componentDidMount() {
    this.routeToMain('Home');
  }

  render() {
    return (
      <View>
        <Text>SIGNIN IN</Text>
      </View>
    );
  }
}
```

**Comment by brentvatne** [comment]
hey @orette - I updated your snack to react-navigation@2.3.1 and the issue was resolved - https://snack.expo.io/BkYm7YeZ7 - this is due to this change https://github.com/react-navigation/react-navigation/releases

**Comment by brentvatne** [comment]
@stychu and others who are having an issue here that isn't resolved in react-navigation@2.3.1: please create a new issue with a runnable mcve (https://stackoverflow.com/help/mcve) and i will be thrilled to help.
